docs/siteconfig.rst
.. _site_config:

Site and Client Configuration
=============================

Teuthology requires several configuration options to be set, and provides many other optional ones. They are looked for in ``~/.teuthology.yaml`` if it exists, or ``/etc/teuthology.yaml`` if it doesn't.

Here is a sample configuration with many of the options set and documented::

    # lab_domain: the domain name to append to all short hostnames
    lab_domain: example.com

    # The root directory to use for storage of all scheduled job logs and 
    # other data.
    archive_base: /home/teuthworker/archive

    # The default machine_type value to use when not specified. Currently 
    # only used by teuthology-suite.
    default_machine_type: awesomebox

    # Control how many machines need to be free in the cluster. 0 means
    # Teuthology can use the entire cluster.
    reserve_machines: 5

    # The machine types currently in active use; currently only used by
    # teuthology-exporter
    active_machine_types: ['smithi']

    # The host and port to use for the beanstalkd queue. This is required 
    # for scheduled jobs.
    queue_host: localhost
    queue_port: 11300

    # The URL of the lock server (paddles). This is required for scheduled 
    # jobs.
    lock_server: http://paddles.example.com:8080/

    # The URL of the results server (paddles).
    results_server: http://paddles.example.com:8080/

    # This URL of the results UI server (pulpito). You must of course use 
    # paddles for pulpito to be useful.
    results_ui_server: http://pulpito.example.com/

    # Email address that will receive job results summaries.
    results_email: ceph-qa@example.com

    # Email address that job results summaries originate from
    results_sending_email: teuthology@example.com

    # How long (in seconds) teuthology-results should wait for jobs to finish
    # before considering them 'hung'
    results_timeout: 43200

    # Gitbuilder archive that stores e.g. ceph packages
    gitbuilder_host: gitbuilder.example.com

    # URL for 'gitserver' helper web application
    # see http://github.com/ceph/gitserver
    githelper_base_url: http://git.ceph.com:8080

    # Verify the packages signatures
    check_package_signatures: true

    # Where all git repos are considered to reside.
    ceph_git_base_url: https://github.com/ceph/

    # Where the ceph git repo is considered to reside.
    ceph_git_url: https://github.com/ceph/ceph.git

    # Where the ceph-qa-suite git repo is considered to reside.
    ceph_qa_suite_git_url: https://github.com/ceph/ceph-qa-suite.git

    # Where teuthology and ceph-qa-suite repos should be stored locally
    src_base_path: /home/foo/src

    # Where teuthology path is located: do not clone if present
    #teuthology_path: .

    # Whether or not teuthology-suite, when scheduling, should update 
    # itself from git. This is disabled by default.
    automated_scheduling: false

    # How often, in seconds, teuthology-worker should poll its child job 
    # processes
    watchdog_interval: 120

    # How long a scheduled job should be allowed to run, in seconds, before 
    # it is killed by the worker process.
    max_job_time: 259200

    # The template from which the URL of the repository containing packages
    # is built.
    #
    # {host} is 'gitbuilder_host' from .teuthology.yaml
    # {proj} is the value of 'project' from the job yaml file or 'ceph'
    # {flavor} is the value of 'flavor' from the job yaml file or 'default'
    # {uri} is ref/tag if 'tag' is set in the job yaml file
    #       or ref/branch if 'branch' is set in the job yaml file
    #       or sha1/sha1 if 'sha1' is set in the job yaml file
    #       or ref/main
    # {pkg_type} is either 'deb' or 'rpm' depending on the host on which the
    #            packages are to be installed
    # {dist} If lsb_release -si is Fedora the value is:
    #          Fedora 20 => fc20
    #          Fedora 21 => fc21
    #          etc.
    #        If lsb_release -si is CentOS or RedHatEnterpriseServer it is
    #          CentOS 6.5 => centos6
    #          CentOS 7.0 => centos7
    #          CentOS 7.1 => centos7
    #          RedHatEnterpriseServer 6.4 => centos6
    #          RedHatEnterpriseServer 7.0 => centos7
    #          RedHatEnterpriseServer 7.1 => centos7
    #          etc.
    #       Everything else is whatever lsb_release -sc returns
    #          Ubuntu 12.04 => precise
    #          Ubuntu 14.04 => trusty
    #          Debian GNU/Linux 7.0 => wheezy
    #          Debian GNU/Linux 8.0 => jessie
    #          etc.
    # {arch} is the output of the 'arch' command on the host on which
    #        the packages are to be installed
    #           i386
    #           x86_64
    #           armv7l
    #           etc.
    baseurl_template: http://{host}/{proj}-{pkg_type}-{dist}-{arch}-{flavor}/{uri}

    # If True, teuthology-suite verifies that a package matching the
    # desired ceph branch exists in the gitbuilder. If False, no
    # verification is done and teuthology-suite assumes the packages
    # are either not necessary to run the task or they are created on
    # demand.
    suite_verify_ceph_hash: True

    # If true, teuthology-suite will schedule jobs even if the required
    # packages are not built.
    suite_allow_missing_packages: False

    # The rsync destination to upload the job results, when --upload is
    # is provided to teuthology-suite.
    #
    archive_upload: ubuntu@teuthology-logs.public.ceph.com:./

    # The path to the SSH private key for rsync to upload to archive_upload
    #
    archive_upload_key: None

    # The public facing URL of the archive_upload location
    #
    archive_upload_url: http://teuthology-logs.public.ceph.com/

    # The OpenStack backend configuration, a dictionary interpreted as follows
    #
    openstack:

      # The teuthology-openstack command will clone teuthology with
      # this command for the purpose of deploying teuthology from
      # scratch and run workers listening on the openstack tube
      #
      clone: git clone http://github.com/ceph/teuthology

      # The path to the user-data file used when creating a target. It can have
      # the {os_type} and {os_version} placeholders which are replaced with
      # the value of --os-type and --os-version. No instance of a give {os_type}
      # and {os_version} combination can be created unless such a file exists.
      #
      user-data: teuthology/openstack/openstack-{os_type}-{os_version}-user-data.txt
  
      # The IP number of the instance running the teuthology cluster. It will
      # be used to build user facing URLs and should usually be the floating IP
      # associated with the instance running the pulpito server.
      #
      ip: 8.4.8.4

      # OpenStack has predefined machine sizes (called flavors)
      # For a given job requiring N machines, the following example select
      # the smallest flavor that satisfies these requirements. For instance
      # If there are three flavors
      #
      #   F1 (10GB disk, 2000MB RAM, 1CPU)
      #   F2 (100GB disk, 7000MB RAM, 1CPU)
      #   F3 (50GB disk, 7000MB RAM, 1CPU)
      #
      # and machine: { disk: 40, ram: 7000, cpus: 1 }, F3 will be chosen.
      # F1 does not have enough RAM (2000 instead of the 7000 minimum) and
      # although F2 satisfies all the requirements, it is larger than F3
      # (100GB instead of 50GB) and presumably more expensive.
      #
      # This configuration applies to all instances created for teuthology jobs
      # that do not redefine these values.
      #
      machine:
      
        # The minimum root disk size of the flavor, in GB
        #
        disk: 20 # GB

        # The minimum RAM size of the flavor, in MB
        # 
        ram: 8000 # MB

        # The minimum number of vCPUS of the flavor
        #
        cpus: 1

      # The volumes attached to each instance. In the following example,
      # three volumes of 10 GB will be created for each instance and 
      # will show as /dev/vdb, /dev/vdc and /dev/vdd
      #
      # This configuration applies to all instances created for teuthology jobs
      # that do not redefine these values.
      #
      volumes:

        # The number of volumes
        #
        count: 3
        
        # The size of each volume, in GB
        #
        size: 10 # GB

    # The host running a [PCP](http://pcp.io/) manager
    pcp_host: http://pcp.front.sepia.ceph.com:44323/

    # Settings for http://www.conserver.com/
    use_conserver: true
    conserver_master: conserver.front.sepia.ceph.com
    conserver_port: 3109

    # Settings for [nsupdate-web](https://github.com/zmc/nsupdate-web)
    # Used by the [libcloud](https://libcloud.apache.org/) backend
    nsupdate_url: http://nsupdate.front.sepia.ceph.com/update

    # Settings for https://fogproject.org/
    fog:
      endpoint: http://fog.example.com/fog
      api_token: your_api_token
      user_token: your_user_token
      machine_types: ['mira', 'smithi']

    # FOG provisioner is default and switching to Pelgas
    # should be made explicitly
    pelagos:
      endpoint: http://head.ses.suse.de:5000/
      machine_types: ['type1', 'type2', 'type3']

    # Do not allow more than that many jobs in a single run by default.
    # To disable this check use 0.
    job_threshold: 500


teuthology/config.py
import os
import yaml
import logging
try:
    from collections.abc import MutableMapping
except ImportError:
    from collections import MutableMapping


def init_logging():
    log = logging.getLogger(__name__)
    return log

log = init_logging()


class YamlConfig(MutableMapping):
    """
    A configuration object populated by parsing a yaml file, with optional
    default values.

    Note that modifying the _defaults attribute of an instance can potentially
    yield confusing results; if you need to do modify defaults, use the class
    variable or create a subclass.
    """
    _defaults = dict()

    def __init__(self, yaml_path=None):
        self.yaml_path = yaml_path
        if self.yaml_path:
            self.load()
        else:
            self._conf = dict()

    def load(self, conf=None):
        if conf:
            if isinstance(conf, dict):
                self._conf = conf
            else:
                self._conf = yaml.safe_load(conf)
            return
        if os.path.exists(self.yaml_path):
            with open(self.yaml_path) as f:
                self._conf = yaml.safe_load(f)
        else:
            log.debug("%s not found", self.yaml_path)
            self._conf = dict()

    def update(self, in_dict):
        """
        Update an existing configuration using dict.update()

        :param in_dict: The dict to use to update
        """
        self._conf.update(in_dict)

    @classmethod
    def from_dict(cls, in_dict):
        """
        Build a config object from a dict.

        :param in_dict: The dict to use
        :returns:       The config object
        """
        conf_obj = cls()
        conf_obj._conf = in_dict
        return conf_obj

    def to_dict(self):
        """
        :returns: A shallow copy of the configuration as a dict
        """
        return dict(self._conf)

    @classmethod
    def from_str(cls, in_str):
        """
        Build a config object from a string or yaml stream.

        :param in_str: The stream or string
        :returns:      The config object
        """
        conf_obj = cls()
        conf_obj._conf = yaml.safe_load(in_str)
        return conf_obj

    def to_str(self):
        """
        :returns: str(self)
        """
        return str(self)

    def get(self, key, default=None):
        return self._conf.get(key, default)

    def __str__(self):
        return yaml.safe_dump(self._conf, default_flow_style=False).strip()

    def __repr__(self):
        return self.__str__()

    def __getitem__(self, name):
        return self.__getattr__(name)

    def __getattr__(self, name):
        return self._conf.get(name, self._defaults.get(name))

    def __contains__(self, name):
        return self._conf.__contains__(name)

    def __setattr__(self, name, value):
        if name.endswith('_conf') or name in ('yaml_path'):
            object.__setattr__(self, name, value)
        else:
            self._conf[name] = value

    def __delattr__(self, name):
        del self._conf[name]

    def __len__(self):
        return self._conf.__len__()

    def __iter__(self):
        return self._conf.__iter__()

    def __setitem__(self, name, value):
        self._conf.__setitem__(name, value)

    def __delitem__(self, name):
        self._conf.__delitem__(name)


class TeuthologyConfig(YamlConfig):
    """
    This class is intended to unify teuthology's many configuration files and
    objects. Currently it serves as a convenient interface to
    ~/.teuthology.yaml and nothing else.
    """
    yaml_path = os.path.join(os.path.expanduser('~/.teuthology.yaml'))
    _defaults = {
        'archive_base': '/home/teuthworker/archive',
        'archive_upload': None,
        'archive_upload_key': None,
        'archive_upload_url': None,
        'automated_scheduling': False,
        'reserve_machines': 5,
        'ceph_git_base_url': 'https://github.com/ceph/',
        'ceph_git_url': None,
        'ceph_qa_suite_git_url': None,
        'ceph_cm_ansible_git_url': None,
        'use_conserver': False,
        'conserver_master': 'conserver.front.sepia.ceph.com',
        'conserver_port': 3109,
        'gitbuilder_host': 'gitbuilder.ceph.com',
        'githelper_base_url': 'http://githelper.ceph.com',
        'check_package_signatures': True,
        'job_threshold': 500,
        'lab_domain': 'front.sepia.ceph.com',
        'lock_server': 'http://paddles.front.sepia.ceph.com/',
        'max_job_time': 259200,  # 3 days
        'nsupdate_url': 'http://nsupdate.front.sepia.ceph.com/update',
        'results_server': 'http://paddles.front.sepia.ceph.com/',
        'results_ui_server': 'http://pulpito.ceph.com/',
        'results_sending_email': 'teuthology',
        'results_timeout': 43200,
        'src_base_path': os.path.expanduser('~/src'),
        'verify_host_keys': True,
        'watchdog_interval': 120,
        'kojihub_url': 'http://koji.fedoraproject.org/kojihub',
        'kojiroot_url': 'http://kojipkgs.fedoraproject.org/packages',
        'koji_task_url': 'https://kojipkgs.fedoraproject.org/work/',
        'baseurl_template': 'http://{host}/{proj}-{pkg_type}-{dist}-{arch}-{flavor}/{uri}',
        'use_shaman': True,
        'shaman_host': 'shaman.ceph.com',
        'teuthology_path': None,
        'suite_verify_ceph_hash': True,
        'suite_allow_missing_packages': False,
        'openstack': {
            'clone': 'git clone http://github.com/ceph/teuthology',
            'user-data': 'teuthology/openstack/openstack-{os_type}-{os_version}-user-data.txt',
            'ip': '1.1.1.1',
            'machine': {
                'disk': 20,
                'ram': 8000,
                'cpus': 1,
            },
            'volumes': {
                'count': 0,
                'size': 1,
            },
        },
        'rocketchat': None,
        'sleep_before_teardown': 0,
        'active_machine_types': [],
    }

    def __init__(self, yaml_path=None):
        super(TeuthologyConfig, self).__init__(yaml_path or self.yaml_path)

    def get_ceph_cm_ansible_git_url(self):
        return (self.ceph_cm_ansible_git_url or
                self.ceph_git_base_url + 'ceph-cm-ansible.git')

    def get_ceph_qa_suite_git_url(self):
        return (self.ceph_qa_suite_git_url or
                self.get_ceph_git_url())

    def get_ceph_git_url(self):
        return (self.ceph_git_url or
                self.ceph_git_base_url + 'ceph-ci.git')


class JobConfig(YamlConfig):
    pass


class FakeNamespace(YamlConfig):
    """
    This class is meant to behave like a argparse Namespace

    We'll use this as a stop-gap as we refactor commands but allow the tasks
    to still be passed a single namespace object for the time being.
    """
    def __init__(self, config_dict=None):
        if not config_dict:
            config_dict = dict()
        self._conf = self._clean_config(config_dict)
        set_config_attr(self)

    def _clean_config(self, config_dict):
        """
        Makes sure that the keys of config_dict are able to be used.  For
        example the "--" prefix of a docopt dict isn't valid and won't populate
        correctly.
        """
        result = dict()
        for key, value in config_dict.items():
            new_key = key
            if new_key.startswith("--"):
                new_key = new_key[2:]
            elif new_key.startswith("<") and new_key.endswith(">"):
                new_key = new_key[1:-1]

            if "-" in new_key:
                new_key = new_key.replace("-", "_")

            result[new_key] = value

        return result

    def __getattr__(self, name):
        """
        We need to modify this for FakeNamespace so that getattr() will
        work correctly on a FakeNamespace instance.
        """
        if name in self._conf:
            return self._conf[name]
        elif name in self._defaults:
            return self._defaults[name]
        raise AttributeError(name)

    def __setattr__(self, name, value):
        if name == 'teuthology_config':
            object.__setattr__(self, name, value)
        else:
            super(FakeNamespace, self).__setattr__(name, value)

    def __repr__(self):
        return repr(self._conf)

    def __str__(self):
        return str(self._conf)


def set_config_attr(obj):
    """
    Set obj.teuthology_config, mimicking the old behavior of misc.read_config
    """
    obj.teuthology_config = config


def _get_config_path():
    system_config_path = '/etc/teuthology.yaml'
    if not os.path.exists(TeuthologyConfig.yaml_path) and \
            os.path.exists(system_config_path):
        return system_config_path
    return TeuthologyConfig.yaml_path

config = TeuthologyConfig(yaml_path=_get_config_path())
